File: content-solutions/binary-search/rotated-shelf.js
function solution(shelf, queries) {
  const n = shelf.length;
  let lo = 0;
  let hi = n - 1;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (shelf[mid] > shelf[n - 1]) lo = mid + 1;
    else hi = mid;
  }
  const p = lo;
  const find = (x, l, h) => {
    while (l <= h) {
      const mid = (l + h) >> 1;
      if (shelf[mid] === x) return mid;
      if (shelf[mid] < x) l = mid + 1;
      else h = mid - 1;
    }
    return -1;
  };
  return queries.map((x) => (x <= shelf[n - 1] ? find(x, p, n - 1) : find(x, 0, p - 1)));
}

File: content-solutions/binary-search/rotated-shelf.java
class Solution {
    int find(int[] a, int x, int lo, int hi) {
        while (lo <= hi) {
            int mid = (lo + hi) >>> 1;
            if (a[mid] == x) return mid;
            if (a[mid] < x) lo = mid + 1;
            else hi = mid - 1;
        }
        return -1;
    }

    public int[] solution(int[] shelf, int[] queries) {
        int n = shelf.length;
        int lo = 0, hi = n - 1;
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (shelf[mid] > shelf[n - 1]) lo = mid + 1;
            else hi = mid;
        }
        int p = lo;
        int[] answer = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int x = queries[i];
            answer[i] = x <= shelf[n - 1] ? find(shelf, x, p, n - 1) : find(shelf, x, 0, p - 1);
        }
        return answer;
    }
}
